src/software/geom/algorithms/end_in_obstacle_sample.cpp

```diff
@@ -0,0 +1,88 @@
+#include "software/geom/algorithms/end_in_obstacle_sample.h"
+
+#include <random>
+
+#include "software/ai/navigator/obstacle/obstacle.hpp"
+#include "software/geom/point.h"
+
+
+std::optional<Point> endInObstacleSample(const std::vector<ObstaclePtr> obstacles,
+                                         const Point &point,
+                                         const Rectangle &navigable_area,
+                                         int initial_count, double rad_step,
+                                         int per_rad_step, double range)
+{
+    bool point_in_obstacle = false;
+    for (auto const &obstacle : obstacles)
+    {
+        if (obstacle->contains(point) || !contains(navigable_area, point))
+        {
+            if (!point_in_obstacle)
+            {
+                point_in_obstacle = true;
+            }
+
+            Point closest_point            = obstacle->closestPoint(point);
+            bool closest_point_in_obstacle = false;
+            for (auto const &obstacle : obstacles)
+            {
+                if (obstacle->contains(closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
+                if (!contains(navigable_area, closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
```

Comment:
Similar to Michael's suggestion, this check doesn't need to be done per obstacle